### PR TITLE
Add MediaTek mt8188 audio support

### PR DESCRIPTION
**Pull request by wenliangwu**
In the pull request, I add mt8188 afe(audio front end) driver for audio control and a simple topology file for mt8188-mt6359 board.

Under drivers/mediatek/afe folder, I create afe folder for audio related driver and sub folders with SoC naming for chip dependent driver to manage common code and chip dependent code. Because memif control driver can be shared by every SoC and it was also implemented as common code in kernel ASoC driver.

### CODE REVIEW

**Comment by dbaluta** [comment]
This looks good AFACT. Please fix some legit checkpatch.pl warnings:

```
- checkpatch wants a different SPDX style for .c versus .h files!
- SOF rejects C99 comments starting with //
```

**Comment by wenliangwu** [comment]
> This looks good AFACT. Please fix some legit checkpatch.pl warnings:
> 
> ```
> - checkpatch wants a different SPDX style for .c versus .h files!
> - SOF rejects C99 comments starting with //
> ```

Hi @dbaluta 

I only see logs as listed below.

> **WARNING: 'AFE' may be misspelled - perhaps 'SAFE'?
> #89: FILE: tools/topology/topology1/sof-mt8188-mt6359.m4:25:
> +# PCM16 ---> AFE (Speaker - rt1019)
>               ^^^
> 
> WARNING: 'AFE' may be misspelled - perhaps 'SAFE'?
> #90: FILE: tools/topology/topology1/sof-mt8188-mt6359.m4:26:
> +# PCM17 ---> AFE (Headset playback - rt5682)
>               ^^^**
> 

How can I find the problem you pointed out?

**Comment by marc-hb** [comment]
`src/platform/mt8188/include/platform/mt8188-afe-common.h` has `// Copyright`, it should be `/* Copyright */`. Some SOF reviewers do not want C99 comments for some unknown reason. I lost track why checkpatch.pl is not catching this, maybe it should (or better: reviewers should stop rejecting C99 comments?)

> - checkpatch wants a different SPDX style for .c versus .h files!

This message is now printed every time there is any checkpatch warning, sorry for any confusion. Please feel free to rephrase it to make that clearer.

Note there are other checkpatch warnings in https://github.com/thesofproject/sof/actions/runs/3937569965/jobs/6735249757 like 
```
WARNING: externs should be avoided in .c files
#89: FILE: src/platform/mt8188/lib/dma.c:20:
+extern struct dma_ops memif_ops;

WARNING: struct dma_ops should normally be const
#89: FILE: src/platform/mt8188/lib/dma.c:20:
+extern struct dma_ops memif_ops;
```

**Comment by wenliangwu** [comment]
> `src/platform/mt8188/include/platform/mt8188-afe-common.h` has `// Copyright`, it should be `/* Copyright */`. Some SOF reviewers do not want C99 comments for some unknown reason. I lost track why checkpatch.pl is not catching this, maybe it should (or better: reviewers should stop rejecting C99 comments?)
> 
Got it. I will modify the `//copyright` to `/* copyright */ `first

> > * checkpatch wants a different SPDX style for .c versus .h files!
> 
> This message is now printed every time there is any checkpatch warning, sorry for any confusion. Please feel free to rephrase it to make that clearer.
> 
Thank you for the information.

> Note there are other checkpatch warnings in https://github.com/thesofproject/sof/actions/runs/3937569965/jobs/6735249757 like
> 
> ```
> WARNING: externs should be avoided in .c files
> #89: FILE: src/platform/mt8188/lib/dma.c:20:
> +extern struct dma_ops memif_ops;
> 
> WARNING: struct dma_ops should normally be const
> #89: FILE: src/platform/mt8188/lib/dma.c:20:
> +extern struct dma_ops memif_ops;
> ```
Thanks, I don't notice it.
I will try fixing these warnings.

**Review by lgirdwood — APPROVED** [review]

**Comment by lgirdwood** [comment]
Code looks fine to me, @wenliangwu pls do work on the checkpatch comments discussed above and I think we are good.

**Review by dbaluta — APPROVED** [review]

File: src/platform/mt8188/afe-platform.c [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+		.rate = 176400,
+		.reg_value = 23,
+	},
+	{
+		.rate = 352800,
+		.reg_value = 24,
+	},
+};
+
+static unsigned int mt8188_afe_fs_timing(unsigned int rate)
```
Comment (yaochunhung) [inline src/platform/mt8188/afe-platform.c:224]: Change to int if you want to return negative value in the end of this function.

File: tools/topology/topology1/sof-mt8188-mt6359.m4 [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+
+dnl PIPELINE_PCM_ADD(pipeline,
+dnl     pipe id, pcm, max channels, format,
+dnl     period, priority, core,
+dnl     pcm_min_rate, pcm_max_rate, pipeline_rate,
+dnl     time_domain, sched_comp)
+
+# Low Latency playback pipeline 1 on PCM 16 using max 2 channels of s16le
+# Set 10000us deadline with priority 0 on core 0
```
Comment (yaochunhung) [inline tools/topology/topology1/sof-mt8188-mt6359.m4:38]: 1000us

File: tools/topology/topology1/sof-mt8188-mt6359.m4 [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+PIPELINE_PCM_ADD(sof/pipe-passthrough-playback.m4,
+	1, 16, 2, s16le,
+	1000, 0, 0,
+	48000, 48000, 48000)
+
+# Low Latency playback pipeline 2 on PCM 17 using max 2 channels of s16le
+# Set 10000us deadline with priority 0 on core 0
```
Comment (yaochunhung) [inline tools/topology/topology1/sof-mt8188-mt6359.m4:45]: 1000us

File: src/platform/mt8188/afe-platform.c (hunk 1 above)
Comment (wenliangwu) [reply src/platform/mt8188/afe-platform.c:225]: Thanks, I will fix it in next version.

Comment (yaochunhung) [reply src/platform/mt8188/afe-platform.c:225]: @wenliangwu . Since it looks like a common issue for mediatek platform. I suggest to commit another PR to fix them. Thank.

Comment (wenliangwu) [reply src/platform/mt8188/afe-platform.c:225]: I found this problem also exists in other platforms. I will create another pull request to fix the problem.

File: tools/topology/topology1/sof-mt8188-mt6359.m4 (hunk 2 above)
Comment (wenliangwu) [reply tools/topology/topology1/sof-mt8188-mt6359.m4]: Done. Thanks.

File: tools/topology/topology1/sof-mt8188-mt6359.m4 (hunk 3 above)
Comment (wenliangwu) [reply tools/topology/topology1/sof-mt8188-mt6359.m4]: Done. Thanks.

**Comment by wenliangwu** [comment]
> Code looks fine to me, @wenliangwu pls do work on the checkpatch comments discussed above and I think we are good.

Thanks, I resolved the comments above. 
Now I only see "afe" related warnings in the checkpatch log.

**Review by yaochunhung — APPROVED** [review]

**Comment by marc-hb** [comment]
> I lost track why checkpatch.pl is not catching this, maybe it should (or better: reviewers should stop rejecting C99 comments?)

Found it while looking for something else:
- https://github.com/thesofproject/sof/pull/5046#issuecomment-999000458

cc: @singalsu

File: src/drivers/mediatek/afe/afe-drv.c [hunk 4]
```diff
@@ -54,29 +55,31 @@ static int afe_memif_set_channel(struct mtk_base_afe *afe, int id, unsigned int
...
 				    memif->data->ch_num_maskbit << memif->data->ch_num_shift,
 				    channel << memif->data->ch_num_shift);
 	}
 
 	if (memif->data->quad_ch_mask) {
-		unsigned int quad_ch = 0;
-
-		if (channel == 4)
-			quad_ch = 1;
+		unsigned int quad_ch = (channel == 4) ? 1 : 0;
```
Comment (lyakh) [inline src/drivers/mediatek/afe/afe-drv.c:65]: all these differences are because you're merging two versions or you're moving and changing?
BTW, this line could just be `quad_ch = channel == 4;`

File: src/platform/mt8188/include/platform/mt8188-afe-reg.h [hunk 5]
```diff
@@ -0,0 +1,2725 @@
...
+#define AFE_CONN134_6                     (0x7218)
+#define AFE_CONN135_6                     (0x721c)
+#define AFE_CONN136_6                     (0x7220)
+#define AFE_CONN137_6                     (0x7224)
+#define AFE_CONN138_6                     (0x7228)
+#define AFE_CONN139_6                     (0x722c)
+#define AFE_CONN176_6                     (0x72c0)
+#define AFE_CONN177_6                     (0x72c4)
+#define AFE_CONN182_6                     (0x72d8)
+#define AFE_CONN183_6                     (0x72dc)
```
[the remark this answers is not in the document]
Comment (wenliangwu) [reply src/platform/mt8188/include/platform/mt8188-afe-reg.h:2723]: Agree. I will remove these parentheses.

**Comment by marc-hb** [comment]
SPDX clarifications submitted in #6973, please review.

**Review by lyakh — APPROVED** [review]

**Review by lgirdwood — APPROVED** [review]
